chain/checkpointing/sub.go

```diff
@@ -0,0 +1,506 @@
+			err = p.UnmarshalBinary(pByte)
+			if err != nil {
+				return nil, err
+			}
+			verificationShares[party.ID(key)] = &p
+		}
+
+		config = &keygen.TaprootConfig{
+			ID:                 party.ID(host.ID().String()),
+			Threshold:          configTOML.Thershold,
+			PrivateShare:       &privateShare,
+			PublicKey:          publickey,
+			VerificationShares: verificationShares,
+		}
+	}
+
+	return &CheckpointingSub{
+		pubsub:    pubsub,
+		topic:     nil,
+		sub:       nil,
+		host:      host,
+		api:       &api,
+		events:    e,
+		init:      false,
+		ptxid:     "",
+		config:    config,
+		newconfig: nil,
+	}, nil
+}
+
+func (c *CheckpointingSub) listenCheckpointEvents(ctx context.Context) {
+
+	checkFunc := func(ctx context.Context, ts *types.TipSet) (done bool, more bool, err error) {
+		return false, true, nil
+	}
+
+	changeHandler := func(oldTs, newTs *types.TipSet, states events.StateChange, curH abi.ChainEpoch) (more bool, err error) {
+		log.Infow("State change detected for power actor")
+
+		idsStrings := c.orderParticipantsList()
+
+		fmt.Println("Participants list :", idsStrings)
+
+		ids := c.formIDSlice(idsStrings)
+
+		id := party.ID(c.host.ID().String())
+
+		threshold := 2
+		n := NewNetwork(ids, c.sub, c.topic)
+		f := frost.KeygenTaproot(id, ids, threshold)
+
+		handler, err := protocol.NewMultiHandler(f, []byte{1, 2, 3})
+		if err != nil {
+			fmt.Println(err)
+			panic(err)
+		}
+		c.LoopHandler(ctx, handler, n)
+		r, err := handler.Result()
+		if err != nil {
+			fmt.Println(err)
+			panic(err)
+		}
+		fmt.Println("Result :", r)
+
+		c.newconfig = r.(*keygen.TaprootConfig)
+
+		return true, nil
+	}
+
+	revertHandler := func(ctx context.Context, ts *types.TipSet) error {
+		return nil
+	}
+
+	match := func(oldTs, newTs *types.TipSet) (bool, events.StateChange, error) {
+		newAct, err := c.api.StateGetActor(ctx, mpower.PowerActorAddr, newTs.Key())
+		if err != nil {
+			return false, nil, err
+		}
+
+		oldAct, err := c.api.StateGetActor(ctx, mpower.PowerActorAddr, oldTs.Key())
+		if err != nil {
+			return false, nil, err
+		}
+
+		var oldSt, newSt mpower.State
+
+		bs := blockstore.NewAPIBlockstore(c.api)
+		cst := cbor.NewCborStore(bs)
+		if err := cst.Get(ctx, oldAct.Head, &oldSt); err != nil {
+			return false, nil, err
+		}
+		if err := cst.Get(ctx, newAct.Head, &newSt); err != nil {
+			return false, nil, err
+		}
+
+		// If Power Actors list has changed start DKG
+		if !c.init {
+			ts, err := c.api.ChainGetTipSetByHeight(ctx, 0, oldTs.Key())
+			if err != nil {
+				panic(err)
+			}
+			data := ts.Cids()[0]
+			err = c.initiate(data.Bytes())
+			if err != nil {
+				panic(err)
+			}
+			return false, nil, nil
+		}
+
+		// ZONDAX TODO
+		// Activate checkpointing every 20 blocks
+		fmt.Println("Height:", newTs.Height())
+		if newTs.Height()%20 == 0 {
+			fmt.Println("Check point time")
+
+			// Initiation and config should be happening at start
+			if c.init && c.config != nil {
+
+				data := oldTs.Cids()[0]
```

Comment:
Quick question (and I apologize in advance if this is obvious but I lack deep knowledge of the protocol): the data included in the checkpoint is just the tipset keys of the epoch when the checkpoint is being committed? Didn't we also add information about the power table in the checkpoint, or that is implicit in the key used to sign?

Reply:
This part of code is outdated. But yes we also have the list of particpants who signed the checkpoint. The latest state of the PR reflect this.